File: evaluation/thumos14/Evaluation/TH14evalDet_Updated.m
function [pr_all,ap_all,map]=TH14evalDet_Updated(detfilename,gtpath,subset,threshold)
% [pr_all,ap_all,map]=TH14evalDet_Updated(detfilename,gtpath,subset,[threshold])
%
%     Input:    detfilename: file path of the input file
%                    gtpath: the path of the groundtruth directory
%                    subset: 'test' or 'val', means the testset or validation set
%                 threshold: overlap threshold (0.5 in default)
%
%    Output:         pr_all: precision-recall curves
%                    ap_all: AP for each class
%                       map: MAP 
%
%
% Evaluation of the temporal detection for 20 classes in the THUMOS 2014 
% action detection challenge http://crcv.ucf.edu/THUMOS14/
%
% The function produces precision-recall curves and average precision
% values for each action class and five values of thresholds for
% the overlap  between ground-truth action intervals and detected action
% intervals. Mean average precision values over classes are also returned.
% 
%
%
% Example:
%
%  [pr_all,ap_all,map]=TH14evalDet('results/Run-2-det.txt','annotation','test',0.5);
%
%
% Plotting precision-recall results:
% 
%  overlapthresh=0.1;
%  ind=find([pr_all.overlapthresh]==overlapthresh);
%  clf
%  for i=1:length(ind)
%    subplot(4,5,i)
%    pr=pr_all(ind(i));
%    plot(pr.rec,pr.prec)
%    axis([0 1 0 1])
%    title(sprintf('%s AP:%1.3f',pr.class,pr.ap))
%  end 
%
  
% THUMOS14 detection classes
%
  
if nargin<4
    threshold=0.5;
end
if nargin<3
    error('At least 3 parameters!')
end


[th14classids,th14classnames]=textread([gtpath '/detclasslist.txt'],'%d%s');
  
% read ground truth
%

clear gtevents
gteventscount=0;
th14classnamesamb=cat(1,th14classnames,'Ambiguous');
for i=1:length(th14classnamesamb)
  class=th14classnamesamb{i};
  gtfilename=[gtpath '/' class '_' subset '.txt'];
  if exist(gtfilename,'file')~=2
    error(['TH14evaldet: Could not find GT file ' gtfilename])
  end
  [videonames,t1,t2]=textread(gtfilename,'%s%f%f');
  for j=1:length(videonames)
    gteventscount=gteventscount+1;
    gtevents(gteventscount).videoname=videonames{j};
    gtevents(gteventscount).timeinterval=[t1(j) t2(j)];
    gtevents(gteventscount).class=class;
    gtevents(gteventscount).conf=1;
  end
end


% parse detection results
%

if exist(detfilename,'file')~=2
  error(['TH14evaldet: Could not find file ' detfilename])
end

[videonames,t1,t2,clsid,conf]=textread(detfilename,'%s%f%f%d%f');
videonames=regexprep(videonames,'\.mp4','');

clear detevents
for i=1:length(videonames)
  ind=find(clsid(i)==th14classids);
  if length(ind)
    detevents(i).videoname=videonames{i};
    detevents(i).timeinterval=[t1(i) t2(i)];
    detevents(i).class=th14classnames{ind};
    detevents(i).conf=conf(i);
  else
    fprintf('WARNING: Reported class ID %d is not among THUMOS14 detection classes.\n')
  end
end

% Evaluate per-class PR for multiple overlap thresholds
%

ap_all=[];
clear pr_all

for i=1:length(th14classnames)
  class=th14classnames{i};
  classid=strmatch(class,th14classnames,'exact');
  assert(length(classid)==1);

    [rec,prec,ap,bgcls_err,othercls_err,loc_err]=TH14eventdetpr(detevents,gtevents,class,threshold);
    pr_all(i,1).class=class;
    pr_all(i,1).classid=classid;
    pr_all(i,1).overlapthresh=threshold;
    pr_all(i,1).prec=prec;
    pr_all(i,1).rec=rec;
    pr_all(i,1).ap=ap;
    pr_all(i,1).bgcls_err=bgcls_err;
    pr_all(i,1).othercls_err=othercls_err;
    pr_all(i,1).loc_err=loc_err;
    ap_all(i,1)=ap;
    
    fprintf('AP:%1.3f at overlap %1.1f for %s\n',ap,threshold,class)
end

map=mean(ap_all,1);
ap_all=ap_all';
fprintf('\n\nMAP: %f \n\n',map);


function [rec,prec,ap,bgcls_err,othercls_err,loc_err]=TH14eventdetpr(detevents,gtevents,class,overlapthresh)

  
gtvideonames={gtevents.videoname};
detvideonames={detevents(:).videoname};
videonames=unique(cat(2,gtvideonames,detvideonames));

%tpconf=[];
%fpconf=[];
unsortConf=[];
unsortFlag=[];
npos=length(strmatch(class,{gtevents.class},'exact'));
assert(npos>0)

indgtclass=strmatch(class,{gtevents.class},'exact');
indambclass=strmatch('Ambiguous',{gtevents.class},'exact');
indothergtclass = setdiff([1:length(gtevents)], [indgtclass;indambclass]);
inddetclass=strmatch(class,{detevents.class},'exact');

if length(inddetclass)==0
    fprintf('Class %s no instance, skip\n',class);
    rec=0;
    prec=0;
    ap=0;
    return;
end

correctPortion=zeros(length(videonames),1);
groundNum=zeros(length(videonames),1);

for i=1:length(videonames)
  correctPortionName{i,1}=videonames{i};
  gt=gtevents(intersect(strmatch(videonames{i},gtvideonames,'exact'),indgtclass));
  othergt=gtevents(intersect(strmatch(videonames{i},gtvideonames,'exact'),indothergtclass));
  amb=gtevents(intersect(strmatch(videonames{i},gtvideonames,'exact'),indambclass)); 
  det=detevents(intersect(strmatch(videonames{i},detvideonames,'exact'),inddetclass));
  
  groundNum(i) = length(gt);
  
  if length(det)
  
    [vs,is]=sort(-[det(:).conf]);
    det=det(is);
    conf=[det(:).conf];
    indfree=ones(1,length(det));
    indamb=zeros(1,length(det));

    % interesct event detection intervals with GT
    if length(gt)
      ov=intervaloverlapvalseconds(cat(1,gt(:).timeinterval),cat(1,det(:).timeinterval));
      % add Loc_Err unmatched
      indfree(sum(ov,1) > 0) = 3;
      for k=1:size(ov,1)
        ind=find(indfree>0);
        [vm,im]=max(ov(k,ind));
        if vm>overlapthresh
            indfree(ind(im))=0;
        end
      end
    end
    
    % interesct event detection intervals with otherGT
    if length(othergt)
      otherov=intervaloverlapvalseconds(cat(1,othergt(:).timeinterval),cat(1,det(:).timeinterval));
      for k=1:size(otherov,1)
        ind=find(indfree==1);
        [vm,im]=max(otherov(k,ind));
        if vm>overlapthresh
            indfree(ind(im))=4;
        end
      end
    end
    
    % respect ambiguous events (overlapping detections will be removed from the FP list)
    if length(amb)
      ovamb=intervaloverlapvalseconds(cat(1,amb(:).timeinterval),cat(1,det(:).timeinterval));
      indamb=sum(ovamb,1);
    end
    
    idx1 = find(indfree==0); %matched
    idx2 = find(indfree==1 & indamb==0); %unmatched BGCls_ERR
    idx3 = find(indfree==3 & indamb==0); %unmatched Loc_ERR
    idx4 = find(indfree==4 & indamb==0); %unmatched OtherCls_ERR

    flag = [ones(size(idx1)) 2*ones(size(idx2)) 3*ones(size(idx3)) 4*ones(size(idx4))];
    [idxall, ttIdx] = sort([idx1 idx2 idx3 idx4]);
    flagall = flag(ttIdx);
    unsortConf = [unsortConf conf(idxall)];
    unsortFlag = [unsortFlag flagall];
    
    %tpconf=[tpconf conf(find(indfree==0))];
    %fpconf=[fpconf conf(find(indfree==1))];
    %fpconf=[fpconf conf(find(indfree==1 & indamb==0))];
    if length(gt)~=0
        correctPortion(i) = length(find(indfree==0))/length(gt);
    end
    
  end
end

%conf=[tpconf fpconf; 1*ones(size(tpconf)) 2*ones(size(fpconf))];
conf=[unsortConf; unsortFlag];

[vs,is]=sort(-conf(1,:));
tp=cumsum(conf(2,is)==1);
fp=cumsum(conf(2,is)==2 | conf(2,is)==3 | conf(2,is)==4);
tmp=conf(2,is)==1;
rec=tp/npos;
prec=tp./(fp+tp);
ap=prap(rec,prec,tmp,npos);
%Cls_Err and Loc_Err
fp_bgcls=cumsum(conf(2,is)==2);
fp_loc=cumsum(conf(2,is)==3);
fp_othercls=cumsum(conf(2,is)==4);

bgcls_err = fp_bgcls ./ (fp+tp) ;
loc_err = fp_loc ./ (fp+tp) ;
othercls_err = fp_othercls ./ (fp+tp) ;
%fprintf('correct: %1.3f,  cls_err: %1.3f, loc_err: %1.3f \n',correct,cls_err,loc_err);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function ap=prap(rec,prec,tmp,npos)

ap=0;
for i=1:length(prec)
    if tmp(i)==1
        ap=ap+prec(i);
    end
end
ap=ap/npos;

File: evaluation/thumos14/Evaluation/python/TH14evalDet_Updated.py
#coding=utf-8

"""
    @lichenyang 2019.10.29

    The python version of TH14evalDet_Updated.m

"""    

import os
import numpy as np
import os.path as osp


from utils import textread1, textread2, textread3, intervaloverlapvalseconds


def TH14evalDet_Updated(detfilename, gtpath, subset, threshold=0.5):
    """
        - outputs:
            pr_all: list[dict]
            ap_all: list
            map: float
    """
    pr_all, ap_all, map = [], [], 0.0

    th14classids, th14classnames = textread2(osp.join(gtpath, 'detclasslist.txt'))

    gtevents = []
    th14classnamesamb = np.append(th14classnames, 'Ambiguous')
    for i in range(len(th14classnamesamb)):
        class_name = th14classnamesamb[i]
        gtfilename = osp.join(gtpath, class_name + '_' + subset + '.txt')
        if not osp.exists(gtfilename):
            print('TH14evaldet: Could not find GT file {}'.format(gtfilename))
            raise
        videonames, t1, t2 = textread3(gtfilename)
        for j in range(len(videonames)):
            gtevent = dict(
                videoname=videonames[j],
                timeinterval=np.array([t1[j], t2[j]]),
                class_name=class_name,
                conf=1,
                )
            gtevents.append(gtevent)


    if not osp.exists(detfilename):
        print('TH14evaldet: Could not find GT file {}'.format(detfilename))
        raise

    videonames, t1, t2, clsid, conf = textread1(detfilename)

    detevents = []
    for i in range(len(videonames)):
        ind = np.nonzero(clsid[i] == th14classids)[0]

        # print(ind)

        if len(ind):
            detevent = dict(
                videoname=videonames[i],
                timeinterval=np.array([t1[i], t2[i]]),
                class_name=th14classnames[ind],
                conf=conf[i],
                )
            detevents.append(detevent)
        else:
            print('WARNING: Reported class ID {} is not \
                among THUMOS14 detection classes.\n'.format(clsid[i]))


    for i in range(len(th14classnames)):
        class_name = th14classnames[i]
        classid = np.where(th14classnames == class_name)[0] + 1 # plus 1: id start from 1
        assert len(classid) == 1

        rec, prec, ap, bgcls_err, othercls_err, loc_err = \
            TH14eventdetpr(detevents, gtevents, class_name, threshold)

        pr_dict = dict(
            class_name=class_name,
            classid=classid,
            overlapthresh=threshold,
            prec=prec,
            rec=rec,
            ap=ap,
            bgcls_err=bgcls_err,
            othercls_err=othercls_err,
            loc_err=loc_err,
            )
        pr_all.append(pr_dict)
        ap_all.append(ap)

        print('AP:{} at overlap {} for {}\n'.format(ap, threshold, class_name))

    map = np.mean(np.array(ap_all))
    print('\n\nMAP: {} \n\n'.format(map))

    return pr_all, ap_all, map



def TH14eventdetpr(detevents, gtevents, class_name, overlapthresh):
    """
        - outputs:
            rec: (N, ), N is the number of detected proposal.
            prec: (N, ), N is the number of detected proposal.
            ap: float
            bgcls_err: (N, ), N is the number of detected proposal.
            othercls_err: (N, ), N is the number of detected proposal.
            loc_err: (N, ), N is the number of detected proposal.
            * Note: N <= len(inddetclass), because some proposals overlap with ambiguous gt are ignored.
    """

    gtvideonames = np.array([item['videoname'] for item in gtevents])
    detvideonames = np.array([item['videoname'] for item in detevents])
    videonames = np.unique(np.concatenate([gtvideonames, detvideonames]))

    # print(len(np.unique(gtvideonames)), len(np.unique(detvideonames)))
    # print(len(videonames))

    unsortConf = np.array([], dtype=np.int)
    unsortFlag = np.array([], dtype=np.int)
    npos = len(np.where(np.array([item['class_name'] for item in gtevents])\
                 == class_name)[0])
    assert npos > 0

    indgtclass = np.where(np.array([item['class_name'] for item in gtevents])\
                 == class_name)[0]
    indambclass = np.where(np.array([item['class_name'] for item in gtevents])\
                 == 'Ambiguous')[0]
    indothergtclass = np.setxor1d(np.arange(len(gtevents)), \
        np.concatenate([indgtclass, indambclass]))
    inddetclass = np.where(np.array([item['class_name'] for item in detevents])\
                 == class_name)[0]

    # print("inddetclass", len(inddetclass))

    if len(inddetclass) == 0:
        print('Class {} no instance, skip\n'.format(class_name))
        return 0,0,0,0,0,0

    correctPortion = np.zeros(len(videonames))
    groundNum = np.zeros(len(videonames))

    for i in range(len(videonames)):

        gt = [gtevents[x] for x in 
            np.intersect1d(np.where(videonames[i] == gtvideonames)[0], indgtclass)]
        othergt = [gtevents[x] for x in 
            np.intersect1d(np.where(videonames[i] == gtvideonames)[0], indothergtclass)]
        amb = [gtevents[x] for x in 
            np.intersect1d(np.where(videonames[i] == gtvideonames)[0], indambclass)]
        det = [detevents[x] for x in 
            np.intersect1d(np.where(videonames[i] == detvideonames)[0], inddetclass)]

        groundNum[i] = len(gt)

        if len(det):

            ids = np.argsort(-np.array([item['conf'] for item in det]))
            det = [det[x] for x in ids]  # descending order
            conf = np.array([item['conf'] for item in det])
            indfree = np.ones(len(det), dtype=np.int)
            indamb = np.zeros(len(det), dtype=np.int)

            # interesct event detection intervals with GT
            if len(gt):
                ov = intervaloverlapvalseconds(
                    np.stack([item['timeinterval'] for item in gt]), 
                    np.stack([item['timeinterval'] for item in det]))
                # print(class_name, videonames[i], ov.shape, ov[0,0])
                indfree[np.sum(ov, axis=0)>0] = 3
                for k in range(ov.shape[0]):
                    ind = np.where(indfree>0)[0]
                    # print(indfree)
                    ov_tmp = ov[k, ind]
                    if len(ov_tmp) == 0:    # len(ov_tmp)==0 means no appropriate proposal.
                        break
                    im = np.argmax(ov_tmp)
                    if ov_tmp[im] > overlapthresh:
                        indfree[ind[im]] = 0

            # interesct event detection intervals with otherGT
            if len(othergt):
                otherov = intervaloverlapvalseconds(
                    np.stack([item['timeinterval'] for item in othergt]), 
                    np.stack([item['timeinterval'] for item in det]))
                for k in range(otherov.shape[0]):
                    ind = np.where(indfree==1)[0]
                    # print(otherov.shape[0],indfree)
                    ov_tmp = otherov[k, ind]
                    if len(ov_tmp) == 0:    # len(ov_tmp)==0 means no appropriate proposal.
                        break
                    im = np.argmax(ov_tmp)
                    if ov_tmp[im] > overlapthresh:
                        indfree[ind[im]] = 4

            # respect ambiguous events (overlapping detections will be removed from the FP list)
            if len(amb):
                ovamb = intervaloverlapvalseconds(
                    np.stack([item['timeinterval'] for item in amb]), 
                    np.stack([item['timeinterval'] for item in det]))
                indamb = np.sum(ovamb, axis=0)


            idx1 = np.where(indfree==0)[0]  # matched
            idx2 = np.where(np.logical_and(indfree==1, indamb==0))[0]  # unmatched BGCls_ERR
            idx3 = np.where(np.logical_and(indfree==3, indamb==0))[0]  # unmatched Loc_ERR
            idx4 = np.where(np.logical_and(indfree==4, indamb==0))[0]  # unmatched OtherCls_ERR

            flag = np.concatenate([
                np.ones(len(idx1)),
                2*np.ones(len(idx2)),
                3*np.ones(len(idx3)),
                4*np.ones(len(idx4)),
                ]).astype(np.int)

            Idx = np.concatenate([idx1, idx2, idx3, idx4])
            ttIdx = np.argsort(Idx) # increasing order
            idxall = Idx[ttIdx]
            # print("idxall", idxall)
            flagall = flag[ttIdx]

            unsortConf = np.append(unsortConf, conf[idxall])    # Actually, conf[idxall] is equal to conf.
            unsortFlag = np.append(unsortFlag, flagall)

            if len(gt) != 0:
                correctPortion[i] = len(np.where(indfree==0)[0]) / len(gt)


    conf = np.stack([unsortConf, unsortFlag])

    # print("conf", conf.shape[1])

    ids = np.argsort(-conf[0, :])   # descending order
    tp = np.cumsum(conf[1, ids]==1)
    fp = np.cumsum(np.logical_or.reduce([conf[1, ids]==2, 
        conf[1, ids]==3, conf[1, ids]==4]))
    tmp = conf[1, ids]==1
    rec = tp / npos
    prec = tp / (fp + tp)
    ap = prap(rec, prec, tmp, npos)

    # Cls_Err and Loc_Err
    fp_bgcls = np.cumsum(conf[1, ids]==2)
    fp_loc = np.cumsum(conf[1, ids]==3)
    fp_othercls = np.cumsum(conf[1, ids]==4)

    bgcls_err = fp_bgcls / (fp+tp)
    loc_err = fp_loc / (fp+tp)
    othercls_err = fp_othercls / (fp+tp)

    return rec, prec, ap, bgcls_err, othercls_err, loc_err



def prap(rec, prec, tmp, npos):
    """
        Calculate the area under the PR curve.
    """
    ap = 0
    for i in range(len(prec)):
        if tmp[i] == 1:
            ap = ap + prec[i]
    ap = ap / npos

    return ap





